packages/terra-date-time-picker/src/DateTimePicker.jsx

```diff
@@ -318,9 +318,57 @@ class DateTimePicker extends React.Component {
 
   handleTimeChange(event, time) {
     this.timeValue = time;
-    const validDate = DateUtil.isValidDate(this.dateValue, this.state.dateFormat) && this.isDateTimeAcceptable(DateTimeUtils.convertDateTimeStringToMomentObject(this.dateValue, this.timeValue, this.state.dateFormat, this.props.showSeconds, this.initialTimeZone));
+    let validDate = DateUtil.isValidDate(this.dateValue, this.state.dateFormat) && this.isDateTimeAcceptable(DateTimeUtils.convertDateTimeStringToMomentObject(this.dateValue, this.timeValue, this.state.dateFormat, this.props.showSeconds, this.initialTimeZone));
     const validTime = DateTimeUtils.isValidTime(this.timeValue, this.props.showSeconds);
-    const previousDateTime = this.state.dateTime ? this.state.dateTime.clone() : null;
+    let previousDateTime = this.state.dateTime ? this.state.dateTime.clone() : null;
+    let dateChanged = false;
+
+    if (event.type === 'keydown') {
+      const timeArray = time.split(':');
+      if (event.key === 'n' || event.key === 'N') {
+        previousDateTime = DateUtil.createSafeDate(DateUtil.getCurrentDate(), this.initialTimeZone);
+        validDate = true;
+        dateChanged = true;
+      }
+
+      if (event.key === '-' || event.key === '_') {
+        if (!validDate) {
+          previousDateTime = DateUtil.createSafeDate(DateUtil.getCurrentDate(), this.initialTimeZone);
+          validDate = true;
+          dateChanged = true;
+        }
+
+        if (timeArray[0] === '23' && timeArray[1] === '59') {
+          previousDateTime = previousDateTime.subtract(1, 'd');
+          dateChanged = true;
+        }
+      }
+
+      if (event.key === '=' || event.key === '+') {
+        if (!validDate) {
+          previousDateTime = DateUtil.createSafeDate(DateUtil.getCurrentDate(), this.initialTimeZone);
+          validDate = true;
+          dateChanged = true;
+        }
+
+        if (timeArray[0] === '00' && timeArray[1] === '00') {
+          previousDateTime = previousDateTime.add(1, 'd');
+          dateChanged = true;
+        }
+      }
+
+      if (this.props.timeVariant === DateTimeUtils.FORMAT_12_HOUR && (event.key === 'a' || event.key === 'A' || event.key === 'p' || event.key === 'P')) {
+        if (!validDate) {
+          previousDateTime = DateUtil.createSafeDate(DateUtil.getCurrentDate(), this.initialTimeZone);
+          validDate = true;
+          dateChanged = true;
+        }
+      }
+    }
+
+    if (dateChanged) {
+      this.handleDateChange(event, previousDateTime.format('YYYY-MM-DD'));
```

Comment:
I don't think we can call handleDateChange here because we already call `this.handleChange` and/or `this.handleChangeRaw` at the end of this function, which will eventually call the `onChange` callback. Call `this.handleDateChange` here again will also eventually call `onChange` callback. As a result, consumers would receive the `onChange` callback twice for every single change.

Reply:
You were correct, and it left over from problem with partial dates, removed here https://github.com/cerner/terra-framework/pull/1520/commits/e31d2250dc60b5bc379214c65e7312ff889cf584#diff-063daa4349be2bbad9c58209d280bfaee7a04b7e429163ebdb7784768f501a5cL369